tensor2tensor/layers/bayes.py

```diff
@@ -98,11 +105,7 @@ def build(self, shape, dtype=None, add_variable_fn=None):
         trainable=True)
 
   def __call__(self):
-    noise = tf.random_normal(self.shape, dtype=self.dtype, seed=self.seed)
-    output = self.mean + self.stddev * noise
-    # TODO(trandustin): Hack to store parameters so KL reg. can operate on them.
-    output._parameters = (self.mean, self.stddev)  # pylint: disable=protected-access
-    return output
+    return ed.Normal(loc=self.mean, scale=self.stddev)
```

Comment:
I was thinking that we could create and store this in the `build` function to remove a bit of Python overhead in eager mode, but it looks like Edward2 RVs only sample from the underlying distribution on the first access in eager mode due to [`rv.value` using a memoization strategy by storing the sample on the first access](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/edward2/random_variable.py#L162).  It seems like it could potentially simplify some use cases if `rv.value` called `rv.distribution.sample` on each access to `rv.value`.  That being said, you definitely thought much longer on the design of Edward[1|2], so I bet there is a reason that would make this undesirable.

Reply:
We mostly memoize the Tensor for usability reasons, where you operate on RandomVariables as if it were any other Tensor. For example, if you call `x = ed.Normal(...)`, it may be less intuitive if `y = x + 1` and `z = x + 1` leads to different results.

On caching: There's an issue that actually arises before Ed2, which is that during `__init__`, Distributions call TF ops on its parameters. This breaks gradients if `Distribution.__init__` isn't recorded in the gradient tape. @vdumoulin raised a point about this in Brain Chat.